packages/flutter/lib/src/material/material.dart

```diff
@@ -548,76 +543,77 @@ class _MaterialState extends State<Material> with TickerProviderStateMixin {
   }
 }
 
-class _RenderInkFeatures extends RenderProxyBox implements MaterialInkController {
-  _RenderInkFeatures({
-    RenderBox? child,
-    required this.vsync,
-    required this.absorbHitTest,
-    this.color,
-  }) : super(child);
+/// A [RenderBox] that paints [Splash]es.
+///
+/// Splash effects can be painted on an ancestor [Material] widget by obtaining
+/// an instance of this controller via [Material.of] and calling [addSplash].
+class SplashController extends RenderProxyBox {
```

Comment:
It would puzzle me as a developer that a controller has methods of render boxes. Is the benefit mostly simplifying writing tests? I suppose the controller is used by end users a lot, and I prefer the previous way for a cleaner set of API.

Reply:
Thank you for explaining the context. But then why should we make it public (other than making tests simpler)? If the developers are not using it, we should just deprecate the public interface after renaming the private class.

Reply:
Good question, thank you.
I believe there are 2 reasons we'd want this to be public:

1. Since **InkWell** is defined in the material library, making **SplashController** private would prevent access.
2. A public class allows for custom **InkWell**/**SplashFactory** implementations.

<br>

I did a [GitHub-wide search](https://github.com/search?q=%22extends+InteractiveInkFeatureFactory%22++NOT+path%3A%22ink_sparkle.dart%22+NOT+path%3A%22ink_splash.dart%22+NOT+path%3A%22ink_ripple.dart%22+NOT+path%3A%22no_splash.dart%22+language%3ADart+&type=code&p=1) for **SplashFactory** subtypes: the vast majority of them had just copy-pasted existing logic with little or no changes. The same was true for [`Material.of(context)`](https://github.com/search?q=%22Material.of%28context%29%22++NOT+path%3A%22ink_well.dart%22+NOT+path%3A%22material.dart%22+NOT+path%3A%22ink_highlight.dart%22+NOT+path%3A%22ink_splash.dart%22+NOT+path%3A%22tabs.dart%22+NOT+path%3A%22text_field.dart%22+language%3ADart+&type=code).

Once we have a **SplashBox** API, perhaps more people will want to try making their own splash factories. But they would run into the exact same problems we have here: the current "public interface/private implementation" setup makes testing a whole lot more difficult.

So even in those cases where someone wants to make their own custom splash effects, I still believe that publicizing the implementation is an overall improvement for the developer experience.

Reply:
Thank you for the explanation. A few more questions:
* Should we name it in a more descriptive way? I mean, for a XxxController I'd expect it to be a mundane class, instead this is a render box. Why didn't we name it `RenderSplash`? (Actually I think I'm hoping we can signify its ability to hold multiple splash effects... but I think that's a topic we've discussed way back and I was probably convinced.)
* Also, have you experimented customizing splashes? Sometimes things that appear customizable don't end up  being so.

Reply:
> Why didn't we name it `RenderSplash`? (Actually I think I'm hoping we can signify its ability to hold multiple splash effects... but I guess that's a topic we've discussed ages ago and I was probably convinced back then.)

That's a really good question—the name `_RenderSplashes` would be analogous to the current `_RenderInkFeatures`.

This comment inspired a fun idea :)

<br>

> Also, have you experimented customizing splashes? Sometimes things that appear customizable don't end up being so.

Fortunately I don't think this will be a problem. The current `splashFactory` options are each defined in their own Dart files, which should hopefully show that a variety of effects can be painted without needing access to any private fields.